Dynamic component with key not compilable anymore

### Marko Version: 4.11.5

### Details

Dynamic components with a `key` attribute lets the compilation fail.

    $ const tag = "h3";
    
    <${tag} key="some-key">
      Hello World!
    </>

But it worked in previous versions of Marko (like 4.7.5)

### Expected Behavior

Compilation should succeed.

### Actual Behavior

Compilation fails.

### Your Environment

Node 8 + Ubuntu 16.04

### Steps to Reproduce

1. install `marko@4.11.5` and `marko-cli`
2. create an `index.marko` file with the template above
3. `marko compile index.marko`

### Stack Trace

```
$ marko compile index.marko
/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:33
        throw createError(
        ^

Error: Unable to compile template at path "/tmp/marko-keyed-dynamic-component/index.marko". Error: tagDef.hasAttribute is not a function
    at transformNode (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:34:13)
    at transformTreeHelper (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:46:5)
    at /tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:57:9
    at ArrayContainer.forEach (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/ast/ArrayContainer.js:18:26)

    at TemplateRoot.forEachChild (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/ast/Node.js:142:23)
    at transformTreeHelper (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:56:10)
    at transformTree (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:75:9)
    at Compiler.compile (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:179:30)
    at _compile (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/index.js:103:33)
    at Object.compileFile (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/index.js:158:16)
Caused by: TypeError: tagDef.hasAttribute is not a function
    at tagDefinitionHasOverridingKeyAttribute (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/components/taglib/components-transformer.js:10:26)
    at Transformer.transform [as _func] (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/components/taglib/components-transformer.js:81:10)
    at /tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:29:33
    at Array.forEach (<anonymous>)
    at TaglibLookup.forEachTagTransformer (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/taglib-lookup/TaglibLookup.js:392:22)
    at TaglibLookup.forEachNodeTransformer (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/taglib-lookup/TaglibLookup.js:341:18)
    at transformNode (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:13:30)
    at transformTreeHelper (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:46:5)
    at /tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/Compiler.js:57:9
    at ArrayContainer.forEach (/tmp/marko-keyed-dynamic-component/node_modules/marko/src/compiler/ast/ArrayContainer.js:18:26)
```

--- a/src/components/taglib/components-transformer.js
+++ b/src/components/taglib/components-transformer.js
@@ -7,7 +7,7 @@ function tagDefinitionHasOverridingKeyAttribute(el) {
     }
 
     var tagDef = el.tagDef;
-    if (tagDef && tagDef.hasAttribute("key")) {
+    if (tagDef && !tagDef.isDynamicTag && tagDef.hasAttribute("key")) {
         return true;
     }
 

--- a/test/components-browser/fixtures/dynamic-tag-user-key/components/hello.marko
+++ b/test/components-browser/fixtures/dynamic-tag-user-key/components/hello.marko
@@ -1,0 +1,6 @@
+class {
+}
+
+<button type="button" on-click('emit', 'hello')>
+    Say hello
+</button>

--- a/test/components-browser/fixtures/dynamic-tag-user-key/index.marko
+++ b/test/components-browser/fixtures/dynamic-tag-user-key/index.marko
@@ -1,0 +1,13 @@
+import helloComponent from './components/hello.marko';
+
+class {
+    onCreate() {
+        this.helloReceived = false;
+    }
+
+    onMount() {
+        this.getComponent('hello').on('hello', () => this.helloReceived = true)
+    }
+}
+
+<${helloComponent} key="hello"/>

--- a/test/components-browser/fixtures/dynamic-tag-user-key/test.js
+++ b/test/components-browser/fixtures/dynamic-tag-user-key/test.js
@@ -1,0 +1,11 @@
+var expect = require("chai").expect;
+
+module.exports = function(helpers) {
+    var component = helpers.mount(require.resolve("./index"), {});
+
+    expect(component.helloReceived).to.equal(false);
+
+    helpers.targetEl.querySelector("button").click();
+
+    expect(component.helloReceived).to.equal(true);
+};

